nemo/lightning/resume.py

```diff
@@ -1,107 +1,81 @@
 from typing import Optional, Union
 
 import lightning_fabric as fl
 import pytorch_lightning as pl
 
 from nemo.lightning import io
-from nemo.lightning.io.mixin import IOMixin
 from nemo.utils import logging
 from nemo.utils.app_state import AppState
 from nemo.utils.model_utils import uninject_model_parallel_rank
 
 # Dynamically inherit from the correct Path subclass based on the operating system.
-if os.name == 'nt':
+if os.name == "nt":
     BasePath = WindowsPath
 else:
     BasePath = PosixPath
 
 
-class Resume(IOMixin):
-    def nemo_path(self, model=None) -> Optional[Path]:
-        """Returns the checkpoint to resume from."""
+@dataclass(kw_only=True)
+class AutoResume:
+    """Class that handles the logic for setting checkpoint paths and restoring from
+    checkpoints in NeMo.
+
+    Attributes:
+        restore_path (str): Path to restore model from. If importing a checkpoint from HF or
+            another non-NeMo checkpoint format, the checkpoint will be automatically converted to a NeMo compatible format.
+            resume_path or the run's log_dir takes precedence over restore_path.
+        resume_path (str): Path to the checkpointing directory to restore from. Defaults to <log_dir>/checkpoints
+        adapter_path (str): Path to any adapter checkpoints.
+        resume_if_exists (bool): Whether this experiment is resuming from a previous run. If
+            True, it sets trainer._checkpoint_connector._ckpt_path so that the trainer should
+            auto-resume. exp_manager will move files under log_dir to log_dir/run_{int}.
+            Defaults to False.
+        resume_past_end (bool): By default, AutoResume throws an error if resume_if_exists is
+            True and a checkpoint matching ``*end.ckpt`` indicating a previous training run
+            fully completed. Setting resume_past_end=True disables this behavior and loads the
+            last checkpoint.
+        resume_ignore_no_checkpoint (bool): AutoResume throws an error if resume_if_exists is
+            True and no checkpoint could be found. Setting resume_ignore_no_checkpoint=True
+            disables this behavior, in which case exp_manager will print a message and
+            continue without restoring.
+        restore_optimizer_states (bool): Flag to control whether to restore optimizer states as well.
+            This has no effect if restore_path is None.
+    """
+
+    restore_path: Optional[str] = None
+    resume_path: Optional[str] = None
+    adapter_path: Optional[str] = None
+    resume_if_exists: bool = False
+    resume_past_end: bool = False
+    resume_ignore_no_checkpoint: bool = False
+    restore_optimizer_states: bool = False
 
     def setup(self, trainer: Union[pl.Trainer, fl.Fabric], model=None):
         if isinstance(trainer, fl.Fabric):
             raise NotImplementedError("Fabric is not supported yet.")
 
-        ckpt_path = self.nemo_path(model)
+        ckpt_path, restore_path = self.nemo_path(model)
```

Comment:
could we rename these to make it clear that one path is for the optimizer and the other is for the model?

Reply:
The `ckpt_path` is for `trainer.ckpt_path` whereas the `restore_path` is for selective restore. I can rename `ckpt_path` to `trainer_ckpt_path`.